// tdesign-react/cjs/dialog/RenderDialog.js
/**
 * tdesign v0.42.2
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../_chunks/dep-3415b40f.js');
var _typeof = require('../_chunks/dep-6c4c2f3e.js');
var slicedToArray = require('../_chunks/dep-8c98b84c.js');
var React = require('react');
var reactTransitionGroup = require('react-transition-group');
var classNames = require('classnames');
var common_Portal = require('../common/Portal.js');
var _util_noop = require('../_util/noop.js');
var _util_useLayoutEffect = require('../_util/useLayoutEffect.js');
var _util_useDialogEsc = require('../_util/useDialogEsc.js');
var dialog_defaultProps = require('./defaultProps.js');
var locale_LocalReceiver = require('../locale/LocalReceiver.js');
require('../_chunks/dep-789d4bfb.js');
require('react-dom');
require('../_util/dom.js');
require('raf');
require('../_chunks/dep-23aa4ed2.js');
require('../_chunks/dep-1ad27019.js');
require('../_chunks/dep-72f2b815.js');
require('../_chunks/dep-3c510ebe.js');
require('../_util/easing.js');
require('../hooks/useConfig.js');
require('../config-provider/ConfigContext.js');
require('../_chunks/dep-bbbce181.js');
require('../_chunks/dep-4c1144f0.js');
require('../_chunks/dep-be113066.js');
require('../_chunks/dep-7affd9c7.js');
require('../_chunks/dep-c15e2539.js');
require('../_chunks/dep-b282bced.js');
require('../_chunks/dep-b8187df5.js');
require('../_chunks/dep-6b7adfaa.js');
require('../_chunks/dep-bdb568fa.js');
require('../_chunks/dep-ffd983b8.js');
require('../_chunks/dep-bafabd9b.js');
require('../_chunks/dep-29ecaf93.js');
require('../_chunks/dep-28bf72fc.js');
require('../_chunks/dep-46b250c3.js');
require('../_chunks/dep-6f0ad8f9.js');
require('../_chunks/dep-901b688e.js');
require('../_chunks/dep-92d05577.js');
require('../_chunks/dep-6aeda844.js');
require('../_chunks/dep-39147135.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../_common/js/global-config/default-config.js');
require('../_chunks/dep-0944d665.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function GetCSSValue(v) {
  return Number.isNaN(Number(v)) ? v : "".concat(Number(v), "px");
}

var transitionTime = 300;
var mousePosition;

var getClickPosition = function getClickPosition(e) {
  mousePosition = {
    x: e.clientX,
    y: e.clientY
  };
  setTimeout(function () {
    mousePosition = null;
  }, 100);
};

if (typeof window !== "undefined" && window.document && window.document.documentElement) {
  document.documentElement.addEventListener("click", getClickPosition, true);
}

var RenderDialog = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("dialog"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 1),
      local = _useLocaleReceiver2[0];

  var prefixCls = props.prefixCls,
      attach = props.attach,
      visible = props.visible,
      mode = props.mode,
      zIndex = props.zIndex,
      showOverlay = props.showOverlay,
      _props$onEscKeydown = props.onEscKeydown,
      onEscKeydown = _props$onEscKeydown === void 0 ? _util_noop["default"] : _props$onEscKeydown,
      _props$onClosed = props.onClosed,
      onClosed = _props$onClosed === void 0 ? _util_noop["default"] : _props$onClosed,
      _props$onClose = props.onClose,
      onClose = _props$onClose === void 0 ? _util_noop["default"] : _props$onClose,
      _props$onCloseBtnClic = props.onCloseBtnClick,
      onCloseBtnClick = _props$onCloseBtnClic === void 0 ? _util_noop["default"] : _props$onCloseBtnClic,
      _props$onOverlayClick = props.onOverlayClick,
      onOverlayClick = _props$onOverlayClick === void 0 ? _util_noop["default"] : _props$onOverlayClick,
      _props$onConfirm = props.onConfirm,
      onConfirm = _props$onConfirm === void 0 ? _util_noop["default"] : _props$onConfirm,
      preventScrollThrough = props.preventScrollThrough,
      closeBtn = props.closeBtn,
      closeOnEscKeydown = props.closeOnEscKeydown,
      confirmOnEnter = props.confirmOnEnter,
      closeOnOverlayClick = props.closeOnOverlayClick,
      destroyOnClose = props.destroyOnClose,
      showInAttachedElement = props.showInAttachedElement;
  var wrap = React.useRef();
  var dialog = React.useRef();
  var dialogPosition = React.useRef();
  var maskRef = React.useRef();
  var portalRef = React.useRef();
  var bodyOverflow = React.useRef();
  var bodyCssTextRef = React.useRef();
  var contentClickRef = React.useRef(false);
  var isModal = mode === "modal";
  var isNormal = mode === "normal";
  var canDraggable = props.draggable && mode === "modeless";
  var dialogOpenClass = "".concat(prefixCls, "__").concat(mode);
  _util_useDialogEsc["default"](visible, wrap);
  React.useImperativeHandle(ref, function () {
    return wrap.current;
  });
  _util_useLayoutEffect["default"](function () {
    bodyOverflow.current = document.body.style.overflow;
    bodyCssTextRef.current = document.body.style.cssText;
  }, []);
  _util_useLayoutEffect["default"](function () {
    if (visible) {
      if (isModal && bodyOverflow.current !== "hidden" && preventScrollThrough && !showInAttachedElement) {
        var scrollWidth = window.innerWidth - document.body.offsetWidth;

        if (bodyCssTextRef.current === "") {
          var bodyCssText = "overflow: hidden;";

          if (scrollWidth > 0) {
            bodyCssText += "position: relative;width: calc(100% - ".concat(scrollWidth, "px);");
          }

          document.body.style.cssText = bodyCssText;
        } else {
          if (scrollWidth > 0) {
            document.body.style.width = "calc(100% - ".concat(scrollWidth, "px)");
            document.body.style.position = "relative";
          }

          document.body.style.overflow = "hidden";
        }
      }

      if (wrap.current) {
        wrap.current.focus();
      }
    } else if (isModal) {
      var openDialogDom = document.querySelectorAll("".concat(prefixCls, "__mode"));

      if (openDialogDom.length < 1) {
        document.body.style.cssText = bodyCssTextRef.current;
      }
    }

    return function () {
      if (isModal) {
        var _openDialogDom = document.querySelectorAll("".concat(prefixCls, "__mode"));

        if (_openDialogDom.length < 1) {
          document.body.style.cssText = bodyCssTextRef.current;
          document.body.style.overflow = bodyOverflow.current;
        }
      } else {
        document.body.style.cssText = bodyCssTextRef.current;
        document.body.style.overflow = bodyOverflow.current;
      }
    };
  }, [preventScrollThrough, attach, visible, mode, isModal, showInAttachedElement, prefixCls]);
  React.useEffect(function () {
    if (visible) {
      if (mousePosition && dialog.current) {
        dialog.current.style.transformOrigin = "".concat(mousePosition.x - dialog.current.offsetLeft, "px ").concat(mousePosition.y - dialog.current.offsetTop, "px");
      }
    }
  }, [visible]);

  var onAnimateLeave = function onAnimateLeave() {
    if (wrap.current) {
      wrap.current.style.display = "none";
    }

    if (isModal && preventScrollThrough) {
      var openDialogDom = document.querySelectorAll("".concat(prefixCls, "__mode"));

      if (isModal && openDialogDom.length < 1) {
        document.body.style.overflow = bodyOverflow.current;
      }
    }

    if (!isModal) {
      var style = dialog.current.style;
      style.position = "relative";
      style.left = "unset";
      style.top = "unset";
    }

    onClosed && onClosed();
  };

  var onMaskClick = function onMaskClick(e) {
    if (showOverlay && (closeOnOverlayClick !== null && closeOnOverlayClick !== void 0 ? closeOnOverlayClick : local.closeOnOverlayClick)) {
      if (contentClickRef.current) {
        contentClickRef.current = false;
      } else if (e.target === dialogPosition.current) {
        onOverlayClick({
          e: e
        });
        onClose({
          e: e,
          trigger: "overlay"
        });
      }
    }
  };

  var handleCloseBtnClick = function handleCloseBtnClick(e) {
    onCloseBtnClick({
      e: e
    });
    onClose({
      e: e,
      trigger: "close-btn"
    });
  };

  var handleKeyDown = function handleKeyDown(e) {
    if (e.key === "Escape") {
      e.stopPropagation();
      onEscKeydown({
        e: e
      });

      if (closeOnEscKeydown !== null && closeOnEscKeydown !== void 0 ? closeOnEscKeydown : local.closeOnEscKeydown) {
        onClose({
          e: e,
          trigger: "esc"
        });
      }
    } else if (e.key === "Enter" || e.key === "NumpadEnter") {
      e.stopPropagation();

      if (confirmOnEnter) {
        onConfirm({
          e: e
        });
      }
    }
  };

  var renderDialog = function renderDialog() {
    var dest = {};

    if (props.width !== void 0) {
      dest.width = GetCSSValue(props.width);
    }

    if (props.mode === "normal") {
      dest.zIndex = "auto";
    }

    var footer = props.footer ? /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(prefixCls, "__footer")
    }, props.footer) : null;
    var header = props.header;
    var body = /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(prefixCls, "__body")
    }, props.body || props.children);
    var closer = closeBtn && /* @__PURE__ */React__default["default"].createElement("span", {
      onClick: handleCloseBtnClick,
      className: "".concat(prefixCls, "__close")
    }, closeBtn);
    var validWindow = (typeof window === "undefined" ? "undefined" : _typeof._typeof(window)) === "object";
    var screenHeight = validWindow ? window.innerHeight || document.documentElement.clientHeight : void 0;
    var screenWidth = validWindow ? window.innerWidth || document.documentElement.clientWidth : void 0;

    var style = _objectSpread(_objectSpread({}, dest), props.style);

    var dialogOffset = {
      x: 0,
      y: 0
    };

    var onDialogMove = function onDialogMove(e) {
      var _dialog$current = dialog.current,
          style2 = _dialog$current.style,
          offsetWidth = _dialog$current.offsetWidth,
          offsetHeight = _dialog$current.offsetHeight;
      var diffX = e.clientX - dialogOffset.x;
      var diffY = e.clientY - dialogOffset.y;
      if (diffX < 0) diffX = 0;
      if (diffY < 0) diffY = 0;
      if (screenWidth - offsetWidth - diffX < 0) diffX = screenWidth - offsetWidth;
      if (screenHeight - offsetHeight - diffY < 0) diffY = screenHeight - offsetHeight;
      style2.position = "absolute";
      style2.left = "".concat(diffX, "px");
      style2.top = "".concat(diffY, "px");
    };

    var onDialogMoveEnd = function onDialogMoveEnd() {
      dialog.current.style.cursor = "default";
      document.removeEventListener("mousemove", onDialogMove);
      document.removeEventListener("mouseup", onDialogMoveEnd);
    };

    var onDialogMoveStart = function onDialogMoveStart(e) {
      contentClickRef.current = true;

      if (canDraggable && e.currentTarget === e.target) {
        var _dialog$current2 = dialog.current,
            offsetLeft = _dialog$current2.offsetLeft,
            offsetTop = _dialog$current2.offsetTop,
            offsetHeight = _dialog$current2.offsetHeight,
            offsetWidth = _dialog$current2.offsetWidth;
        if (offsetWidth > screenWidth || offsetHeight > screenHeight) return;
        dialog.current.style.cursor = "move";
        var diffX = e.clientX - offsetLeft;
        var diffY = e.clientY - offsetTop;
        dialogOffset = {
          x: diffX,
          y: diffY
        };
        document.addEventListener("mousemove", onDialogMove);
        document.addEventListener("mouseup", onDialogMoveEnd);
      }
    };

    var positionStyle = {};

    if (props.top) {
      var topValue = GetCSSValue(props.top);
      positionStyle.paddingTop = topValue;
    }

    var positionClass = classNames__default["default"]("".concat(prefixCls, "__position"), defineProperty._defineProperty({}, "".concat(prefixCls, "--top"), !!props.top), "".concat(props.placement && !props.top ? "".concat(prefixCls, "--").concat(props.placement) : ""));
    var dialogElement = /* @__PURE__ */React__default["default"].createElement("div", {
      className: isNormal ? "" : "".concat(prefixCls, "__wrap")
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: isNormal ? "" : positionClass,
      style: positionStyle,
      onClick: onMaskClick,
      ref: dialogPosition
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      ref: dialog,
      style: style,
      className: classNames__default["default"]("".concat(prefixCls), "".concat(prefixCls, "--default")),
      onMouseDown: onDialogMoveStart
    }, closer, header, body, footer)));
    return /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.CSSTransition, {
      "in": props.visible,
      appear: true,
      mountOnEnter: true,
      unmountOnExit: destroyOnClose,
      timeout: transitionTime,
      classNames: "".concat(prefixCls, "-zoom"),
      onEntered: props.onOpened,
      onExited: onAnimateLeave,
      nodeRef: dialog
    }, dialogElement);
  };

  var renderMask = function renderMask() {
    var maskElement;

    if (showOverlay) {
      maskElement = /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.CSSTransition, {
        "in": visible,
        appear: true,
        timeout: transitionTime,
        classNames: "".concat(prefixCls, "-fade"),
        mountOnEnter: true,
        unmountOnExit: true,
        nodeRef: maskRef
      }, /* @__PURE__ */React__default["default"].createElement("div", {
        ref: maskRef,
        className: "".concat(prefixCls, "__mask")
      }));
    }

    return maskElement;
  };

  var render = function render() {
    var style = {};

    if (visible) {
      style.display = "block";
    }

    var wrapStyle = _objectSpread(_objectSpread({}, style), {}, {
      zIndex: zIndex
    });

    var dialogBody = renderDialog();
    var wrapClass = classNames__default["default"](props.className, "".concat(prefixCls, "__ctx"), !isNormal ? "".concat(prefixCls, "__ctx--fixed") : "", visible ? dialogOpenClass : "", isModal && showInAttachedElement ? "".concat(prefixCls, "__ctx--absolute") : "", props.mode === "modeless" ? "".concat(prefixCls, "__ctx--modeless") : "");
    var dialog2 = /* @__PURE__ */React__default["default"].createElement("div", {
      ref: wrap,
      className: wrapClass,
      style: wrapStyle,
      onKeyDown: handleKeyDown,
      tabIndex: 0
    }, mode === "modal" && renderMask(), dialogBody);
    var dom = null;

    if (visible || wrap.current) {
      if (attach === "" || isNormal) {
        dom = dialog2;
      } else {
        dom = /* @__PURE__ */React__default["default"].createElement(reactTransitionGroup.CSSTransition, {
          "in": visible,
          appear: true,
          timeout: transitionTime,
          mountOnEnter: true,
          unmountOnExit: destroyOnClose,
          nodeRef: portalRef
        }, /* @__PURE__ */React__default["default"].createElement(common_Portal["default"], {
          attach: attach,
          ref: portalRef
        }, dialog2));
      }
    }

    return dom;
  };

  return render();
});
RenderDialog.defaultProps = dialog_defaultProps.dialogDefaultProps;

exports["default"] = RenderDialog;
