// tdesign-react/lib/jumper/Jumper.js
/**
 * tdesign v0.42.2
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-ebef14de.js';
import React, { forwardRef, useMemo } from 'react';
import classNames from 'classnames';
import { i as isObject_1 } from '../_chunks/dep-056b9d81.js';
import { RoundIcon, ChevronUpIcon, ChevronDownIcon, ChevronLeftIcon, ChevronRightIcon } from 'tdesign-icons-react';
import useConfig from '../hooks/useConfig.js';
import useGlobalIcon from '../hooks/useGlobalIcon.js';
import { Button } from '../button/index.js';
import { jumperDefaultProps } from './defaultProps.js';
import noop from '../_util/noop.js';
import '../_chunks/dep-3627969a.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-98b209dd.js';
import '../_chunks/dep-1304d812.js';
import '../_chunks/dep-339b0951.js';
import '../_chunks/dep-ba4fa05d.js';
import '../_chunks/dep-00ad7a91.js';
import '../_chunks/dep-cf61d761.js';
import '../_chunks/dep-86599c44.js';
import '../_chunks/dep-de91e914.js';
import '../_chunks/dep-f60756e1.js';
import '../_chunks/dep-090fa3b2.js';
import '../_chunks/dep-a98a054e.js';
import '../_chunks/dep-7a725039.js';
import '../_chunks/dep-e48ce47b.js';
import '../_chunks/dep-d9495851.js';
import '../_chunks/dep-c393cab2.js';
import '../_chunks/dep-75ee7646.js';
import '../_chunks/dep-68ef3987.js';
import '../_chunks/dep-7237da98.js';
import '../_chunks/dep-d905b9e9.js';
import '../_chunks/dep-87e9108c.js';
import '../_chunks/dep-873eecd0.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import '../button/Button.js';
import '../_chunks/dep-fe234c83.js';
import '../_util/useRipple.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-f3228585.js';
import '../_util/easing.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../_chunks/dep-2baf8638.js';
import '../_chunks/dep-24599104.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-88be14a4.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../button/defaultProps.js';

var Jumper = /*#__PURE__*/forwardRef(function (props, ref) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useGlobalIcon = useGlobalIcon({
    RoundIcon: RoundIcon,
    ChevronUpIcon: ChevronUpIcon,
    ChevronDownIcon: ChevronDownIcon,
    ChevronLeftIcon: ChevronLeftIcon,
    ChevronRightIcon: ChevronRightIcon
  }),
      RoundIcon$1 = _useGlobalIcon.RoundIcon,
      ChevronUpIcon$1 = _useGlobalIcon.ChevronUpIcon,
      ChevronDownIcon$1 = _useGlobalIcon.ChevronDownIcon,
      ChevronLeftIcon$1 = _useGlobalIcon.ChevronLeftIcon,
      ChevronRightIcon$1 = _useGlobalIcon.ChevronRightIcon;

  var variant = props.variant,
      tips = props.tips,
      showCurrent = props.showCurrent,
      disabled = props.disabled,
      layout = props.layout,
      size = props.size,
      _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? noop : _props$onChange,
      className = props.className,
      style = props.style;
  var titleConfig = useMemo(function () {
    if (isObject_1(tips)) return tips;
    if (tips === true) return {
      prev: "\u4E0A\u4E00\u9875",
      current: "\u5F53\u524D",
      next: "\u4E0B\u4E00\u9875"
    };
    return {};
  }, [tips]);
  var disabledConfig = useMemo(function () {
    if (isObject_1(disabled)) return disabled;
    if (disabled === true) return {
      prev: true,
      current: true,
      next: true
    };
    return {
      prev: false,
      current: false,
      next: false
    };
  }, [disabled]);
  return /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(classPrefix, "-jumper"), className, _defineProperty({}, "".concat(classPrefix, "-jumper--outline"), variant === "outline")),
    ref: ref,
    style: style
  }, /* @__PURE__ */React.createElement(Button, {
    title: titleConfig.prev,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "prev"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React.createElement(ChevronLeftIcon$1, null) : /* @__PURE__ */React.createElement(ChevronUpIcon$1, null),
    className: "".concat(classPrefix, "-jumper__prev"),
    disabled: disabledConfig.prev
  }), showCurrent && /* @__PURE__ */React.createElement(Button, {
    title: titleConfig.current,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "current"
      });
    },
    icon: /* @__PURE__ */React.createElement(RoundIcon$1, null),
    className: "".concat(classPrefix, "-jumper__current"),
    disabled: disabledConfig.current
  }), /* @__PURE__ */React.createElement(Button, {
    title: titleConfig.next,
    variant: variant,
    size: size,
    shape: "square",
    onClick: function onClick(e) {
      return onChange({
        e: e,
        trigger: "next"
      });
    },
    icon: layout === "horizontal" ? /* @__PURE__ */React.createElement(ChevronRightIcon$1, null) : /* @__PURE__ */React.createElement(ChevronDownIcon$1, null),
    className: "".concat(classPrefix, "-jumper__next"),
    disabled: disabledConfig.next
  }));
});
Jumper.displayName = "Jumper";
Jumper.defaultProps = jumperDefaultProps;

export { Jumper as default };
