// tdesign-react/esm/date-picker/panel/PanelContent.js
/**
 * tdesign v0.42.2
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-ebef14de.js';
import React from 'react';
import DatePickerHeader from '../base/Header.js';
import DatePickerTable from '../base/Table.js';
import { TimePickerPanel } from '../../time-picker/index.js';
import useConfig from '../../hooks/useConfig.js';
import { getDefaultFormat } from '../../_common/js/date-picker/format.js';
import '../../_chunks/dep-88be14a4.js';
import '../../_chunks/dep-24599104.js';
import '../../_chunks/dep-2baf8638.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../_chunks/dep-a407cb09.js';
import '../../_chunks/dep-588efa4a.js';
import '../../_chunks/dep-dd4cd0af.js';
import '../../_chunks/dep-f162740d.js';
import '../../_chunks/dep-22eee49a.js';
import '../../_chunks/dep-f2b0265c.js';
import '../../_chunks/dep-91563b2d.js';
import '../../_chunks/dep-3627969a.js';
import '../../_chunks/dep-e10ccd97.js';
import '../../_chunks/dep-6f568a94.js';
import '../../_chunks/dep-646bb331.js';
import '../../_chunks/dep-1f71931a.js';
import '../../_chunks/dep-3b5c7604.js';
import '../../_chunks/dep-51566efd.js';
import '../../_chunks/dep-c6a2f197.js';
import '../../_chunks/dep-a6b91546.js';
import '../../_chunks/dep-5fef4495.js';
import '../../_chunks/dep-a97b932f.js';
import '../../_chunks/dep-9e2d72a2.js';
import '../../_chunks/dep-342e02ed.js';
import '../../_chunks/dep-a89633b7.js';
import '../../_chunks/dep-a91fe24d.js';
import '../../_chunks/dep-27e7ca24.js';
import '../../_chunks/dep-f40ed698.js';
import '../../_chunks/dep-ab50f53f.js';
import '../../_chunks/dep-9703c0d5.js';
import '../../config-provider/ConfigContext.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../select/index.js';
import '../../select/base/Select.js';
import '../../_chunks/dep-fe234c83.js';
import 'classnames';
import '../../_chunks/dep-eaafd9b2.js';
import '../../_chunks/dep-41762eda.js';
import '../../_chunks/dep-1a0fcc8a.js';
import '../../_chunks/dep-8e8d0d95.js';
import '../../hooks/useControlled.js';
import '../../_chunks/dep-6f296cb1.js';
import '../../_chunks/dep-8b68327e.js';
import '../../_chunks/dep-094f627b.js';
import '../../_util/noop.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../select/util/helper.js';
import '../../select/base/OptionGroup.js';
import '../../select/defaultProps.js';
import '../../select/base/Option.js';
import '../../_chunks/dep-4c44a6fb.js';
import '../../_chunks/dep-2968ca73.js';
import '../../_util/useRipple.js';
import '../../_util/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';
import '../../common/FakeArrow.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';
import '../../loading/style/index.js';
import '../../loading/type.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../popup/utils/ref.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-622322f2.js';
import '../../_chunks/dep-abf310d4.js';
import '../../_util/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/style/index.js';
import '../../popup/type.js';
import '../../select-input/useSingle.js';
import '../../_chunks/dep-17055df6.js';
import '../../_chunks/dep-81e269a3.js';
import '../../_chunks/dep-80d1a65c.js';
import '../../_chunks/dep-9a198ec7.js';
import '../../input/index.js';
import '../../input/Input.js';
import 'tdesign-icons-react';
import '../../_util/useLayoutEffect.js';
import '../../hooks/useGlobalIcon.js';
import '../../input/InputGroup.js';
import '../../input/defaultProps.js';
import '../../input/style/index.js';
import '../../input/type.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import '../../_util/useDragSorter.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/useTagList.js';
import '../../tag/index.js';
import '../../tag/Tag.js';
import '../../tag/CheckTag.js';
import '../../tag/defaultProps.js';
import '../../tag/style/index.js';
import '../../tag/type.js';
import '../../tag-input/useHover.js';
import '../../tag-input/defaultProps.js';
import '../../tag-input/style/index.js';
import '../../tag-input/type.js';
import '../../select-input/useOverlayInnerStyle.js';
import '../../select-input/defaultProps.js';
import '../../select-input/style/index.js';
import '../../select-input/type.js';
import '../../select/base/PopupContent.js';
import '../../select/style/index.js';
import '../../select/type.js';
import '../../jumper/index.js';
import '../../jumper/Jumper.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../button/defaultProps.js';
import '../../button/style/index.js';
import '../../button/type.js';
import '../../jumper/defaultProps.js';
import '../../jumper/style/index.js';
import '../../jumper/type.js';
import '../base/Cell.js';
import '../../_common/js/date-picker/utils.js';
import 'dayjs';
import '../../_chunks/dep-e23ef400.js';
import '../../time-picker/TimePicker.js';
import '../../time-picker/TimeRangePicker.js';
import '../../range-input/index.js';
import '../../range-input/RangeInput.js';
import '../../range-input/defaultProps.js';
import '../../range-input/RangeInputPopup.js';
import '../../range-input/style/index.js';
import '../../range-input/type.js';
import '../../time-picker/panel/TimePickerPanel.js';
import '../../time-picker/panel/SinglePanel.js';
import '../../time-picker/hooks/useTimePickerTextConfig.js';
import '../../_common/js/time-picker/const.js';
import '../../_common/js/time-picker/utils.js';
import '../../hooks/useDebounce.js';
import '../../_util/usePersistFn.js';
import '../../hooks/usePropsRef.js';
import '../../time-picker/defaultProps.js';
import '../../time-picker/style/index.js';
import '../../time-picker/type.js';
import '../../_common/js/log/index.js';
import '../../_common/js/log/log.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function PanelContent(props) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var panelName = "".concat(classPrefix, "-date-picker__panel");
  var value = props.value,
      year = props.year,
      month = props.month,
      mode = props.mode,
      format = props.format,
      enableTimePicker = props.enableTimePicker,
      timePickerProps = props.timePickerProps,
      firstDayOfWeek = props.firstDayOfWeek,
      _props$partial = props.partial,
      partial = _props$partial === void 0 ? "start" : _props$partial,
      time = props.time,
      tableData = props.tableData,
      _onMonthChange = props.onMonthChange,
      _onYearChange = props.onYearChange,
      _onJumperClick = props.onJumperClick,
      _onCellClick = props.onCellClick,
      _onCellMouseEnter = props.onCellMouseEnter,
      onCellMouseLeave = props.onCellMouseLeave,
      onTimePickerChange = props.onTimePickerChange;

  var _getDefaultFormat = getDefaultFormat({
    mode: mode,
    format: format,
    enableTimePicker: enableTimePicker
  }),
      timeFormat = _getDefaultFormat.timeFormat;

  var showTimePicker = enableTimePicker && mode === "date";
  var defaultTime = "00:00:00";
  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(panelName, "-content")
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(panelName, "-").concat(mode)
  }, /* @__PURE__ */React.createElement(DatePickerHeader, {
    mode: mode,
    year: year,
    month: month,
    onMonthChange: function onMonthChange(val) {
      return _onMonthChange === null || _onMonthChange === void 0 ? void 0 : _onMonthChange(val, {
        partial: partial
      });
    },
    onYearChange: function onYearChange(val) {
      return _onYearChange === null || _onYearChange === void 0 ? void 0 : _onYearChange(val, {
        partial: partial
      });
    },
    onJumperClick: function onJumperClick(_ref) {
      var trigger = _ref.trigger;
      return _onJumperClick === null || _onJumperClick === void 0 ? void 0 : _onJumperClick({
        trigger: trigger,
        partial: partial
      });
    }
  }), /* @__PURE__ */React.createElement(DatePickerTable, {
    mode: mode,
    data: tableData,
    value: value,
    time: time,
    format: format,
    firstDayOfWeek: firstDayOfWeek,
    onCellClick: function onCellClick(date, _ref2) {
      var e = _ref2.e;
      return _onCellClick === null || _onCellClick === void 0 ? void 0 : _onCellClick(date, {
        e: e,
        partial: partial
      });
    },
    onCellMouseEnter: function onCellMouseEnter(date) {
      return _onCellMouseEnter === null || _onCellMouseEnter === void 0 ? void 0 : _onCellMouseEnter(date, {
        partial: partial
      });
    },
    onCellMouseLeave: onCellMouseLeave
  })), showTimePicker && /* @__PURE__ */React.createElement("div", {
    className: "".concat(panelName, "-time")
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(panelName, "-time-viewer")
  }, time || defaultTime), /* @__PURE__ */React.createElement(TimePickerPanel, _objectSpread({
    key: partial,
    format: timeFormat,
    value: time || defaultTime,
    onChange: onTimePickerChange,
    isShowPanel: props.popupVisible
  }, timePickerProps))));
}

export { PanelContent as default };
