// tdesign-react/lib/table/EnhancedTable.js
/**
 * tdesign v0.42.2
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-ebef14de.js';
import { _ as _objectWithoutProperties } from '../_chunks/dep-fe234c83.js';
import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import PrimaryTable from './PrimaryTable.js';
import useTreeData from './hooks/useTreeData.js';
import useTreeSelect from './hooks/useTreeSelect.js';
import '../_chunks/dep-bca278f2.js';
import '../_chunks/dep-6a110e92.js';
import '../_chunks/dep-c393cab2.js';
import '../_chunks/dep-3627969a.js';
import '../_chunks/dep-893e0247.js';
import '../_chunks/dep-cf61d761.js';
import '../_chunks/dep-090fa3b2.js';
import '../_chunks/dep-339b0951.js';
import '../_chunks/dep-ba4fa05d.js';
import '../_chunks/dep-00ad7a91.js';
import '../_chunks/dep-056b9d81.js';
import '../_chunks/dep-86599c44.js';
import '../_chunks/dep-2998e451.js';
import 'classnames';
import './BaseTable.js';
import '../_chunks/dep-2baf8638.js';
import '../_chunks/dep-24599104.js';
import '../_chunks/dep-e5ce5161.js';
import '../_chunks/dep-fd4264c2.js';
import '../_chunks/dep-7a725039.js';
import '../_chunks/dep-e48ce47b.js';
import '../_chunks/dep-75ee7646.js';
import '../_chunks/dep-d9495851.js';
import '../_chunks/dep-f60756e1.js';
import '../_chunks/dep-e65dabd7.js';
import '../_chunks/dep-de8845e2.js';
import '../_chunks/dep-d905b9e9.js';
import './TBody.js';
import '../_chunks/dep-d24ebdc1.js';
import '../_chunks/dep-8f4dca99.js';
import '../_chunks/dep-65354ae3.js';
import '../_chunks/dep-4192679d.js';
import '../_chunks/dep-504b652f.js';
import './TR.js';
import './utils.js';
import './hooks/useFixed.js';
import '../_common/js/utils/helper.js';
import '../_chunks/dep-88be14a4.js';
import '../_common/js/log/log.js';
import '../_common/js/utils/getScrollbarWidth.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-f3228585.js';
import '../_util/easing.js';
import '../hooks/useDebounce.js';
import '../_chunks/dep-1997a475.js';
import '../_chunks/dep-d502af5f.js';
import '../_util/usePersistFn.js';
import './hooks/useClassName.js';
import '../hooks/useConfig.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-98b209dd.js';
import '../_chunks/dep-1304d812.js';
import '../_chunks/dep-de91e914.js';
import '../_chunks/dep-a98a054e.js';
import '../_chunks/dep-68ef3987.js';
import '../_chunks/dep-7237da98.js';
import '../_chunks/dep-87e9108c.js';
import '../_chunks/dep-873eecd0.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_common/js/global-config/default-config.js';
import './Ellipsis.js';
import '../tooltip/index.js';
import '../tooltip/Tooltip.js';
import '../popup/index.js';
import '../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../hooks/useControlled.js';
import '../_util/noop.js';
import '../_util/useAnimation.js';
import '../common/Portal.js';
import 'react-dom';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../tooltip/defaultProps.js';
import '../tooltip/TooltipLite.js';
import '../_util/useSwitch.js';
import '../_common/js/utils/getPosition.js';
import './hooks/useLazyLoad.js';
import '../_common/js/utils/observe.js';
import './hooks/useRowspanAndColspan.js';
import '../locale/LocalReceiver.js';
import '../affix/index.js';
import '../affix/Affix.js';
import '../affix/defaultProps.js';
import './THead.js';
import './hooks/useTableHeader.js';
import './hooks/useMultiHeader.js';
import './TFoot.js';
import './hooks/useColumnResize.js';
import '../_chunks/dep-04fa7e7c.js';
import '../_common/js/table/set-column-width-by-drag.js';
import '../_common/js/table/utils.js';
import '../_common/js/table/recalculate-column-width.js';
import './hooks/useAffix.js';
import './hooks/usePagination.js';
import '../pagination/index.js';
import '../pagination/Pagination.js';
import '../_chunks/dep-5374504a.js';
import '../_chunks/dep-3a12b5c0.js';
import '../_chunks/dep-37dd0f2e.js';
import '../select/index.js';
import '../select/base/Select.js';
import '../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../select/util/helper.js';
import '../select/base/OptionGroup.js';
import '../select/defaultProps.js';
import '../select/base/Option.js';
import '../_util/useRipple.js';
import '../_util/setStyle.js';
import '../common/FakeArrow.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import '../select-input/index.js';
import '../select-input/SelectInput.js';
import '../select-input/useSingle.js';
import '../input/index.js';
import '../input/Input.js';
import 'tdesign-icons-react';
import '../_util/useLayoutEffect.js';
import '../hooks/useGlobalIcon.js';
import '../input/InputGroup.js';
import '../input/defaultProps.js';
import '../select-input/useMultiple.js';
import '../tag-input/index.js';
import '../tag-input/TagInput.js';
import '../_util/useDragSorter.js';
import '../tag-input/useTagScroll.js';
import '../tag-input/useTagList.js';
import '../tag/index.js';
import '../tag/Tag.js';
import '../tag/CheckTag.js';
import '../tag/defaultProps.js';
import '../tag-input/useHover.js';
import '../tag-input/defaultProps.js';
import '../select-input/useOverlayInnerStyle.js';
import '../select-input/defaultProps.js';
import '../select/base/PopupContent.js';
import '../input-number/index.js';
import '../input-number/InputNumber.js';
import '../button/index.js';
import '../button/Button.js';
import '../button/defaultProps.js';
import '../input-number/useInputNumber.js';
import '../_util/useCommonClassName.js';
import '../_common/js/input-number/number.js';
import '../_common/js/input-number/large-number.js';
import '../input-number/defaultProps.js';
import '../input-adornment/InputAdornment.js';
import '../pagination/hooks/useBoundaryJumper.js';
import '../pagination/hooks/usePrevNextJumper.js';
import '../pagination/hooks/usePageNumber.js';
import '../pagination/hooks/useTotal.js';
import '../pagination/validators.js';
import '../pagination/defaultProps.js';
import './hooks/useStyle.js';
import '../hooks/useCommonClassName.js';
import './defaultProps.js';
import './hooks/useColumnController.js';
import '../_chunks/dep-11f8d73a.js';
import '../_chunks/dep-b7bdd272.js';
import '../checkbox/index.js';
import '../_chunks/dep-27763e39.js';
import '../common/Check.js';
import '../_util/helper.js';
import '../checkbox/defaultProps.js';
import '../dialog/plugin.js';
import '../dialog/Dialog.js';
import '../dialog/RenderDialog.js';
import '../_util/useDialogEsc.js';
import '../dialog/defaultProps.js';
import '../_util/useSetState.js';
import './hooks/useRowExpand.js';
import './hooks/useRowSelect.js';
import '../radio/index.js';
import '../_chunks/dep-52579c55.js';
import '../radio/defaultProps.js';
import './hooks/useSorter.js';
import './SorterButton.js';
import './hooks/useFilter.js';
import './FilterController.js';
import '../_chunks/dep-fbe9bd73.js';
import './hooks/useDragSort.js';
import 'sortablejs';
import '../_common/js/utils/swapDragArrayElement.js';
import './hooks/useAsyncLoading.js';
import './EditableCell.js';
import '../form/formModel.js';
import '../_chunks/dep-18bd07e0.js';
import './hooks/useEditableRow.js';
import '../_chunks/dep-46dc5ea7.js';
import '../_common/js/table/tree-store.js';
import '../_chunks/dep-a58d2703.js';

var _excluded = ["store", "dataSource", "formatTreeColumn", "swapData"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var EnhancedTable = /*#__PURE__*/forwardRef(function (props, ref) {
  var tree = props.tree,
      columns = props.columns,
      style = props.style,
      className = props.className;
  var primaryTableRef = useRef();

  var _useTreeData = useTreeData(props),
      store = _useTreeData.store,
      dataSource = _useTreeData.dataSource,
      formatTreeColumn = _useTreeData.formatTreeColumn,
      swapData = _useTreeData.swapData,
      treeInstanceFunctions = _objectWithoutProperties(_useTreeData, _excluded);

  var treeDataMap = store === null || store === void 0 ? void 0 : store.treeDataMap;

  var _useTreeSelect = useTreeSelect(props, treeDataMap),
      tIndeterminateSelectedRowKeys = _useTreeSelect.tIndeterminateSelectedRowKeys,
      onInnerSelectChange = _useTreeSelect.onInnerSelectChange;

  var getColumns = function getColumns(columns2) {
    var arr = [];

    for (var i = 0, len = columns2.length; i < len; i++) {
      var _item$children, _item$children2;

      var item = _objectSpread({}, columns2[i]);

      item = formatTreeColumn(item);

      if ((_item$children = item.children) !== null && _item$children !== void 0 && _item$children.length) {
        item.children = getColumns(item.children);
      }

      if (!item.children || (_item$children2 = item.children) !== null && _item$children2 !== void 0 && _item$children2.length) {
        arr.push(item);
      }
    }

    return arr;
  };

  var tColumns = function () {
    var isTreeData = !tree || !Object.keys(tree).length;
    return isTreeData ? columns : getColumns(columns);
  }();

  useImperativeHandle(ref, function () {
    return _objectSpread(_objectSpread({
      treeDataMap: treeDataMap
    }, treeInstanceFunctions), primaryTableRef.current);
  });

  var onDragSortChange = function onDragSortChange(params) {
    var _props$onDragSort;

    if (props.beforeDragSort && !props.beforeDragSort(params)) return;
    swapData({
      data: params.data,
      current: params.current,
      target: params.target,
      currentIndex: params.currentIndex,
      targetIndex: params.targetIndex
    });
    (_props$onDragSort = props.onDragSort) === null || _props$onDragSort === void 0 ? void 0 : _props$onDragSort.call(props, params);
  };

  var primaryTableProps = _objectSpread(_objectSpread({}, props), {}, {
    data: dataSource,
    columns: tColumns,
    indeterminateSelectedRowKeys: tIndeterminateSelectedRowKeys,
    disableDataPage: Boolean(tree && Object.keys(tree).length),
    onSelectChange: onInnerSelectChange,
    onDragSort: onDragSortChange,
    style: style,
    className: className
  });

  return /* @__PURE__ */React.createElement(PrimaryTable, _objectSpread(_objectSpread({}, primaryTableProps), {}, {
    ref: primaryTableRef
  }));
});
EnhancedTable.displayName = "EnhancedTable";

export { EnhancedTable as default };
