// tdesign-react/esm/table/hooks/useTableHeader.js
/**
 * tdesign v0.42.2
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _typeof } from '../../_chunks/dep-3627969a.js';
import { _ as _slicedToArray } from '../../_chunks/dep-2baf8638.js';
import { _ as _defineProperty } from '../../_chunks/dep-ebef14de.js';
import React, { useMemo } from 'react';
import { i as isFunction_1 } from '../../_chunks/dep-f2b0265c.js';
import classNames from 'classnames';
import { getThRowspanAndColspan, getThList } from './useMultiHeader.js';
import useClassName from './useClassName.js';
import Ellipsis from '../Ellipsis.js';
import '../../_chunks/dep-24599104.js';
import '../../_chunks/dep-a407cb09.js';
import '../../_chunks/dep-91563b2d.js';
import '../../_chunks/dep-e10ccd97.js';
import '../../_chunks/dep-6f568a94.js';
import '../../hooks/useConfig.js';
import '../../config-provider/ConfigContext.js';
import '../../_chunks/dep-588efa4a.js';
import '../../_chunks/dep-dd4cd0af.js';
import '../../_chunks/dep-f162740d.js';
import '../../_chunks/dep-22eee49a.js';
import '../../_chunks/dep-646bb331.js';
import '../../_chunks/dep-1f71931a.js';
import '../../_chunks/dep-3b5c7604.js';
import '../../_chunks/dep-51566efd.js';
import '../../_chunks/dep-c6a2f197.js';
import '../../_chunks/dep-a6b91546.js';
import '../../_chunks/dep-5fef4495.js';
import '../../_chunks/dep-a97b932f.js';
import '../../_chunks/dep-9e2d72a2.js';
import '../../_chunks/dep-342e02ed.js';
import '../../_chunks/dep-a89633b7.js';
import '../../_chunks/dep-a91fe24d.js';
import '../../_chunks/dep-27e7ca24.js';
import '../../_chunks/dep-f40ed698.js';
import '../../_chunks/dep-ab50f53f.js';
import '../../_chunks/dep-9703c0d5.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_common/js/global-config/default-config.js';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-2968ca73.js';
import '../../_util/easing.js';
import '../../tooltip/index.js';
import '../../tooltip/Tooltip.js';
import '../../_chunks/dep-fe234c83.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../hooks/useControlled.js';
import '../../_chunks/dep-6f296cb1.js';
import '../../_chunks/dep-8b68327e.js';
import '../../_chunks/dep-094f627b.js';
import '../../_chunks/dep-8e8d0d95.js';
import '../../_chunks/dep-1a0fcc8a.js';
import '../../_util/noop.js';
import '../../_util/useAnimation.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../popup/hooks/useTrigger.js';
import 'react-is';
import '../../popup/utils/ref.js';
import '../../_util/composeRefs.js';
import '../../popup/utils/transition.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-622322f2.js';
import '../../_chunks/dep-abf310d4.js';
import '../../_util/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/style/index.js';
import '../../popup/type.js';
import '../../tooltip/defaultProps.js';
import '../../tooltip/TooltipLite.js';
import '../../_util/useSwitch.js';
import '../../_util/usePersistFn.js';
import '../../_common/js/utils/getPosition.js';
import '../../tooltip/style/index.js';
import '../../tooltip/type.js';
import '../../hooks/useDebounce.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function renderTitle(col, index) {
  var params = {
    col: col,
    colIndex: index
  };

  if (isFunction_1(col.title)) {
    return col.title(params);
  }

  if (isFunction_1(col.render)) {
    return col.render(_objectSpread(_objectSpread({}, params), {}, {
      row: {},
      rowIndex: -1,
      type: "title"
    }));
  }

  return col.title;
}
function useTableHeader(_ref) {
  var columns = _ref.columns;

  var _useClassName = useClassName(),
      tableSortClasses = _useClassName.tableSortClasses,
      tableFilterClasses = _useClassName.tableFilterClasses;

  var spansAndLeafNodes = useMemo(function () {
    return getThRowspanAndColspan(columns);
  }, [columns]);
  var thList = useMemo(function () {
    return getThList(columns);
  }, [columns]);
  var isMultipleHeader = useMemo(function () {
    return thList.length > 1;
  }, [thList]);

  var renderTitleWidthIcon = function renderTitleWidthIcon(_ref2, col, colIndex, ellipsisTitle, attach, extra) {
    var _classes;

    var _ref3 = _slicedToArray(_ref2, 3),
        title = _ref3[0],
        sortIcon = _ref3[1],
        filterIcon = _ref3[2];

    var classes = (_classes = {}, _defineProperty(_classes, tableSortClasses.sortable, !!sortIcon), _defineProperty(_classes, tableFilterClasses.filterable, !!filterIcon), _classes);
    var content = isFunction_1(ellipsisTitle) ? ellipsisTitle({
      col: col,
      colIndex: colIndex
    }) : void 0;
    var isEllipsis = ellipsisTitle !== void 0 ? Boolean(ellipsisTitle) : Boolean(col.ellipsis);
    return /* @__PURE__ */React.createElement("div", {
      className: classNames(classes)
    }, /* @__PURE__ */React.createElement("div", {
      className: tableSortClasses.title
    }, isEllipsis ? /* @__PURE__ */React.createElement(Ellipsis, {
      placement: "bottom",
      attach: attach ? function () {
        return attach;
      } : void 0,
      popupContent: content,
      tooltipProps: _typeof(ellipsisTitle) === "object" ? ellipsisTitle : void 0,
      classPrefix: extra === null || extra === void 0 ? void 0 : extra.classPrefix,
      overlayClassName: extra === null || extra === void 0 ? void 0 : extra.ellipsisOverlayClassName
    }, title) : /* @__PURE__ */React.createElement("div", null, title), Boolean(sortIcon || filterIcon) && /* @__PURE__ */React.createElement("div", {
      className: tableFilterClasses.iconWrap
    }, sortIcon, filterIcon)));
  };

  return {
    thList: thList,
    isMultipleHeader: isMultipleHeader,
    spansAndLeafNodes: spansAndLeafNodes,
    renderTitleWidthIcon: renderTitleWidthIcon
  };
}

export { useTableHeader as default, renderTitle };
